contracts/vesting.token/src/vesting.token.cpp

```diff
@@ -0,0 +1,72 @@
+#include <vesting.token/vesting.token.hpp>
+
+namespace vestingtoken {
+    void vestingToken::updatedate(eosio::time_point_sec newStartDate) {
+        require_auth(get_self());
+        startDate startDate(get_self(), get_self().value);
+        startDate.set(newStartDate, get_self());
+    }
+    void vestingToken::assigntokens(eosio::name holder, eosio::asset amount, VestingCategory category) {
+        // Only the contract owner can call this function
+        require_auth(get_self());
+
+        // Check the symbol is correct and valid
+        auto sym = amount.symbol;
+        eosio::check(sym == system_resource_currency, "Symbol does not match system resource currency");
+        eosio::check(sym.precision() == system_resource_currency.precision(), "Symbol precision does not match");
+        eosio::check(amount.amount > 0, "Amount must be greater than 0");
+
+        // Create a new vesting schedule
+        vesting_schedules vestingTable(get_self(), holder.value);
+        auto iter = vestingTable.end();
+
+        vestingTable.emplace(get_self(), [&](auto& row) {
+            row.holder = holder;
+            row.totalTokens = amount;
+            row.tokensClaimed = asset(0, amount.symbol);
+            row.allocated = eosio::current_time_point();
+            row.vestingCategory = category;
+        });
+    }
+    void vestingToken::withdraw(eosio::name holder){
+        require_auth(holder);
+        // Get the vesting schedule
+        vesting_schedules vestingTable(get_self(), holder.value);
+        for (auto iter = vestingTable.begin(); iter != vestingTable.end(); ++iter) {
+            const vested_allocation& vestingSchedule = *iter;
+
+            VestingCategory category = vestingSchedule.vestingCategory;
+
+            eosio::time_point_sec startDelayDays = eosio::time_point_sec(category.startDelayDays * 24 * 60 * 60);
+            eosio::time_point_sec startDateValue = startDate(get_self(), get_self().value).get();
+            eosio::time_point_sec vesting_start = startDateValue + eosio::seconds(startDelayDays.sec_since_epoch()) + vestingSchedule.allocated;
+
+
+            // Calculate the tokens that can be claimed
+            if (eosio::current_time_point() > eosio::time_point(vesting_start)) {
+                eosio::asset cliffTokens = vestingSchedule.totalTokens * category.cliffPeriodDays /
+                                        category.vestingPeriodDays;
+                
+                eosio::microseconds microseconds_elapsed = eosio::current_time_point() - vesting_start;
```

Comment:
this is microseconds, other units use seconds. I don't think this is correct due to this.

Reply:
eosio::current_time_point() this function returns the current time as a eosio::time_point object
it represents the number of microseconds

Reply:
see above. lets just keep seconds and days, not also microseconds. Also, we DO need to multiply/divide by 1000 to convert from microseconds to seconds. The constructors do not do this for you, they just expect a number as you can see here https://docs.antelope.io/cdt/latest/reference/Classes/classeosio_1_1microseconds#public-functions